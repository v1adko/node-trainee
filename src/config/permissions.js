export default {
  PUBLIC: { role: 'public', value: 0 },
  USER: { role: 'user', value: 1 },
  ADMIN: { role: 'admin', value: 2 }
};
