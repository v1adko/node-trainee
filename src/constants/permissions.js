export default {
  PUBLIC: { value: 'public', priority: 0 },
  USER: { value: 'user', priority: 1 },
  MODERATOR: { value: 'moderator', priority: 2 },
  ADMIN: { value: 'admin', priority: 3 }
};
