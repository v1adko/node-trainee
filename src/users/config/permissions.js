export default {
  CREATE_USER: 'createUser',
  READ_USER: 'readUser',
  UPDATE_USER: 'updateUser',
  DELETE_USER: 'deleteUser',
  GET_MY_DATE: 'getMyDate',
  UPDATE_MY_DATE: 'updateMyDate'
};
