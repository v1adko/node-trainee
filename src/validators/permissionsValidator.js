import permissionsConst from '../config/permissions';

function getRoleValue(role) {
  return permissionsConst[role.toUpperCase()].value;
}

function permissionsValidator(permissions) {
  return (req, res, next) => {
    if (permissions.value >= getRoleValue(req.user.role)) {
      return next();
    }
    return res
      .status(403)
      .send({ message: "You don't have permission for this action" });
  };
}

export default permissionsValidator;
