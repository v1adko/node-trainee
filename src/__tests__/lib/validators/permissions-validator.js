import permissionsValidator from '../../../lib/validators/permissions-validator';
import permissions from '../../../constants/permissions';

const request = {
  user: {
    role: permissions.USER.value
  }
};

describe('Test permissions validator', async () => {
  it('should check permissions, nothing to do and give pass to next methods', async () => {
    permissionsValidator(permissions.USER, request);
  });

  it('should return PermissionsError, because permissions does not enough', async () => {
    const exceedAccess = () => permissionsValidator(permissions.ADMIN, request);

    expect(exceedAccess).toThrowErrorMatchingSnapshot();
  });

  it('should return PermissionsError, because request does not have permissions', async () => {
    const exceedAccess = () => permissionsValidator(permissions.ADMIN, {});

    expect(exceedAccess).toThrowErrorMatchingSnapshot();
  });

  it('should pass all methods, because permission level does not specified', async () => {
    permissionsValidator({}, request);
  });
});
