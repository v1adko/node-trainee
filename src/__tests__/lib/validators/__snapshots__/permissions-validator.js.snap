// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test permissions validator should return PermissionsError, because permissions does not enough 1`] = `"Access was denied. Not enough permissions."`;

exports[`Test permissions validator should return PermissionsError, because request does not have permissions 1`] = `"Access was denied. Not enough permissions."`;
